inspirehep/modules/records/api.py

```diff
@@ -95,6 +152,243 @@ def validate(self):
+            key = self._get_unique_files_key(base_file_name=file_name)
+
+        if stream is not None:
+            self.files[key] = stream
+
+        builder = LiteratureBuilder(record=self.dumps())
+        metadata['key'] = key
+        metadata['url'] = '/api/files/{bucket}/{key}'.format(
+            bucket=self.files[key].bucket_id,
+            key=key,
+        )
+        if is_document:
+            builder.add_document(**metadata)
+        else:
+            builder.add_figure(**metadata)
+
+        super(InspireRecord, self).update(builder.record)
+        return metadata
+
+    def _resolve_doc_or_fig_url(
+        self,
+        doc_or_fig_obj,
+        src_record=None,
+        only_new=False,
+    ):
+        """Resolves the given document url according to the current record.
+
+        It will fill it up with the local url in case it's found in the given
+        src_record.
+        It will leave any document that does not need to be downloaded with the
+        url intact, pointing to the local files api endpoint.
+        """
+        doc_or_fig_obj = copy.deepcopy(doc_or_fig_obj)
+        key = doc_or_fig_obj['key']
+
+        if not doc_or_fig_obj['url'].startswith('/api/files/'):
+            return doc_or_fig_obj
+
+        if not src_record:
+            if not only_new:
+                raise Exception(
+                    "Can't download %s, as no src_record was passed."
+                    % doc_or_fig_obj
+                )
+            else:
+                if key not in self.files:
+                    raise Exception(
+                        "Bad document %s, refers to an already downloaded "
+                        "url, but none found in this record and no src_record "
+                        "passed, only found %s."
+                        % (doc_or_fig_obj, self.files.keys)
+                    )
+
+                else:
+                    return doc_or_fig_obj
+
+        else:
+            if key not in src_record.files and key not in self.files:
+                raise Exception(
+                    "Bad document %s, refers to an already downloaded "
+                    "url, but none found in this record or the passed "
+                    "src_record, only found %s on this record and %s on "
+                    "src_record." % (
+                        doc_or_fig_obj,
+                        self.files.keys,
+                        src_record.files.keys,
+                    )
+                )
+            if key not in self.files:
+                doc_or_fig_obj['url'] = src_record.files[key].file.uri
+
+        return doc_or_fig_obj
+
+    def _download_to_docs_or_figs(
+        self,
+        document=None,
+        figure=None,
+        src_record=None,
+        only_new=False,
+    ):
+        if not document and not figure:
+            raise TypeError(
+                'No document nor figure passed, at least one is needed.'
+            )
+
+        is_document = bool(document)
```

Comment:
Just wondering if this is enough? What does a document look like?

Reply:
It's a dictionary with some keys, it's defined in the schemas, I just rely in the fact that there's at least on key required. Should be enough.